router/dataplane_spec.gobra

```diff
@@ -268,3 +270,542 @@ func closureSpec2(i uint16, c BatchConn)
+	0 < getSeg1Len(hdr) &&
+	0 <= getSeg2Len(hdr) && 
+	0 <= getSeg3Len(hdr) &&
+	4 + 8 * numInfoFields(int(getSeg1Len(hdr)), int(getSeg2Len(hdr)), int(getSeg3Len(hdr))) + 12 * (int(getSeg1Len(hdr)) + int(getSeg2Len(hdr)) + int(getSeg3Len(hdr))) == len(raw)
+decreases 
+pure func getPkt(dp io.DataPlaneSpec, raw []byte, asid io.IO_as) option[io.IO_pkt2]{
+	return let hdr := unfolding acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _) in binary.BigEndian.Uint32(raw[0:4]) in
+		let currINFIdx := int(getCurrINF(hdr)) in
+		let currHFIdx := int(getCurrHF(hdr)) in
+		let seg1Len := int(getSeg1Len(hdr)) in
+		let seg2Len := int(getSeg2Len(hdr)) in 
+		let seg3Len := int(getSeg3Len(hdr)) in
+		let segLen := lengthOfCurrSeg(currHFIdx, seg1Len, seg2Len, seg3Len) in
+		let prevSegLen := lengthOfPrevSeg(currHFIdx, seg1Len, seg2Len, seg3Len) in
+		let numINF := int(numInfoFields(seg1Len, seg2Len, seg3Len)) in
+		let offset := 4 + 8 * numINF in
+		let consDir := unfolding acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _) in getConsDir(raw, currINFIdx) in
+		let currAsidSeq := getAsidSeq(dp, raw, numINF, currHFIdx, prevSegLen+segLen, prevSegLen, consDir, dp.Asid()) in
+		currAsidSeq == none[seq[io.IO_as]] ? none[io.IO_pkt2] :
+			let last := get(currAsidSeq)[segLen-1] in
+			let first := get(currAsidSeq)[0] in
+			let leftAsidSeq := getAsidsLeft(dp, raw, numINF, currINFIdx + 1, seg1Len, seg2Len, seg3Len, last) in
+			let rightAsidSeq := getAsidsRight(dp, raw, numINF, currINFIdx - 1, seg1Len, seg2Len, seg3Len, first) in
+			(leftAsidSeq == none[seq[io.IO_as]] || rightAsidSeq == none[seq[io.IO_as]]) ? none[io.IO_pkt2] :
+				let midAsid := ((currINFIdx == 0 && seg2Len > 0 && seg3Len > 0) ? some(get(leftAsidSeq)[len(get(leftAsidSeq))-1]) :
+					(currINFIdx == 2 && seg2Len > 0) ? some(get(rightAsidSeq)[0]) : none[io.IO_as]) in
+				let midAsidSeq := getAsidsMid(dp, raw, numINF, currINFIdx + 2, seg1Len, seg2Len, seg3Len, midAsid) in
+				midAsidSeq == none[seq[io.IO_as]] ? none[io.IO_pkt2] :
+					some(io.IO_pkt2(io.IO_Packet2{
+						CurrSeg : getCurrSeg(raw, offset+prevSegLen, currINFIdx, currHFIdx-prevSegLen, segLen, get(currAsidSeq)),
+						LeftSeg : getLeftSeg(raw, offset, currINFIdx + 1, seg1Len, seg2Len , seg3Len, get(leftAsidSeq)),
+						MidSeg : getMidSeg(raw, offset, currINFIdx + 2, seg1Len, seg2Len , seg3Len, get(midAsidSeq)),
+						RightSeg : getRightSeg(raw, offset, currINFIdx - 1, seg1Len, seg2Len , seg3Len, get(rightAsidSeq)),
+					}))
+}
+
+ghost
+decreases
+pure func getCurrINF(hdr uint32) uint8{
+	return uint8(hdr >> 30)
+}
+
+ghost
+decreases
+pure func getCurrHF(hdr uint32) uint8{
+	return uint8(hdr >> 24) & 0x3F
+}
+
+ghost
+decreases
+pure func getSeg1Len(hdr uint32) uint8{
+	return uint8(hdr >> 12) & 0x3F
+}
+
+ghost
+decreases
+pure func getSeg2Len(hdr uint32) uint8{
+	return uint8(hdr >> 6) & 0x3F
+}
+
+ghost
+decreases
+pure func getSeg3Len(hdr uint32) uint8{
+	return uint8(hdr) & 0x3F
+}
```

Comment:
Don't we have anything like that in the package `slayers/path`? If so, please use those instead. Otherwise, maybe it should exist there (or rather, in the package where we parse these fields from the Header, instead of this package - in a sense, we are exposing implementation details from some package, thus it should be in that package. Also, In general, I fell like functions named `get...` are very unidiomatic in Go. I would call these `CurrHF`, `LenSeg1`, `LenSeg2`, ...